Feature Alternatives to sudo for privilege escalation

TheSemicolons opened the pull request:
### What does this PR do?
This commit adds functionality for alternative privilege escalation
commands to be used via the `sudo_cmd` variable in the minion config
file. If the variable is absent the default of `sudo` is used.


### What issues does this PR fix or reference?
Fixes #67255 

### Previous Behavior
Only `sudo` was available for privilege escalation.

### New Behavior
`sudo` is the default privilege escalation. `sudo_cmd` has been added
and allows alternatives to be specified in the minion config file via this
variable.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the test documentation for details on how to implement tests
into Salt's test suite:
https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

twangboy reviewed (changes requested):
We probably want to add this as a new config option by adding it to ``salt/config/__init__.py`` and document it in ``conf/minion`` and ``doc/ref/configuration/minion.rst``

twangboy reviewed (changes requested):
This also needs a changelog and a test

twangboy reviewed (changes requested):
Now all we need is a test

dwoz committed:
Feature: add sudo_cmd minion config option for alternative privilege escalation

Adds `sudo_cmd` minion config option to allow alternatives to `sudo`
(e.g. `doas`) for privilege escalation when `sudo_user` is configured.

- Add `sudo_cmd` to VALID_OPTS and DEFAULT_MINION_OPTS in salt/config/__init__.py
- Update `salt/executors/sudo.py` __virtual__ and execute() to use `sudo_cmd`
- Document `sudo_cmd` in conf/minion and doc/ref/configuration/minion.rst
- Add comprehensive unit tests covering default sudo, sudo_cmd alternative,
  --priv arg, and __virtual__ load/fail scenarios
- Add changelog entry for issue #67255

Fixes #67255

twangboy commented on doc/ref/configuration/minion.rst:
```diff
@@ -564,6 +564,20 @@ need to be changed to the ownership of the new user.
+.. code-block:: yaml
+
+    sudo_cmd: doas
+
```
Extra white space

twangboy commented on salt/executors/sudo.py:
```diff
@@ -12,7 +12,9 @@
 
 def __virtual__():
-    if salt.utils.path.which("sudo") and __opts__.get("sudo_user"):
+    if salt.utils.path.which(__opts__.get("sudo_cmd") or "sudo") and __opts__.get(
+        "sudo_user"
+    ):
```
If a user explicitly sets sudo_cmd: /path/to/broken_or_missing_command in their minion config, salt.utils.path.which() will return None. Because the condition fails, __virtual__ returns False, and the entire executor module fails to load.

```python
def __virtual__():
    # Fall back to "sudo" if sudo_cmd is empty/unset
    cmd = __opts__.get("sudo_cmd") or "sudo"
    
    # Ensure either the configured command OR the fallback "sudo" is available
    if (salt.utils.path.which(cmd) or salt.utils.path.which("sudo")) and __opts__.get("sudo_user"):
        return __virtualname__
    return False
```

twangboy commented on tests/pytests/unit/executors/test_sudo.py:
```diff
@@ -30,23 +29,102 @@ def test_sudo_execute_adds_priv_arg():
+    opts = {"sudo_user": "saltdev", "sudo_cmd": "doas"}
+    with _with_opts(opts), patch("salt.utils.path.which", return_value=None):
+        result = salt.executors.sudo.__virtual__()
+    assert result is False
```
Once you make the change to salt.executors.sudo, you'll want to add a test case in tests/pytests/unit/executors/test_sudo.py to ensure that __virtual__ still returns "sudo" if sudo_cmd is broken but sudo is present on the minion:

```python
def test_sudo_virtual_loads_with_fallback_when_sudo_cmd_invalid(tmp_path):
    """__virtual__ returns virtualname if explicit sudo_cmd is missing but sudo exists."""
    fake_sudo = tmp_path / "sudo"
    fake_sudo.touch()
    fake_sudo.chmod(0o755)
    
    opts = { "sudo_user": "saltdev", "sudo_cmd": "broken_command" }
    with _with_opts(opts), patch(
        "salt.utils.path.which",
        side_effect=lambda x: str(fake_sudo) if x == "sudo" else None,
    ):
        result = salt.executors.sudo.__virtual__()
        assert result == "sudo"
```
